### ISSUE TITLE
7 Year limit on due dates

### ISSUE THREAD

**Issue by mhughes2k** [jmcgettrick/MoodleDirectV2#277]
Following a support call with TII, we discovered that there is a 7 year limit on due dates.

We discovered this after our event queue got blocked, as the TII implementation simply bails out, instead of catching this, which appears in the Activity log as 

```
2015-02-25 16:43:45 +0000 (API_ERROR) - There was an error trying to create the assignment in Turnitin<br/><br/>Fault Code: invaliddata | File: /moodle/web/mod/turnitintooltwo/sdk/assignmentsoap.class.php | Line: 33 | Message: date_due - cannot exceed the term length of the class
 | Code: 0^M
```

This *shouldn't* break the event queue, but should warn the user in some shape or form. 

The plagiarism plugin form add an additional validation rule to the "due date" element that kicks in if the  due date is set to far in advance and the plagiarism plugin is turned on (admittedly this would then *prevent* the activity being set up until set to a "sensible" date).

**Comment by jmcgettrick** [comment]
Hi Michael, this is the first time we've came across this error in the plugin. We'll look at adding validation for the due date as you've suggested which should get round this. We'll keep you posted

**Comment by amclarty** [comment]
Michael, out of interest, what was the admin/instructor trying to achieve with an assignment with a due date set 7 years in the future? Are we missing a common scenario that long running assignments are used for?

**Comment by mhughes2k** [comment]
We have an MBA programme that is quite flexible, and the course staff were "experimenting" in a test class.

Unfortunately it then blocked the event queue for the entire server :-(

It's more of a "defensive" approach, because (IMHO) setting a due date definitely shouldn't cause everything to stop working.

I do think that ideally this needs a 2 pronged approach:
1. the turnitin code should be catching this error and ideally handling it rather than  reporting the event as failed. As the sevice has made a valid response (i.e. this date is too far in the future).
2 the users should be given some feedback about it occurring.

So an variation/extension to the previous suggestion would be that:
1. A static / html element is inserted by the PP after the duedate element, and if the due date is too far in advance, this displays the information to the user, along the lines of "Turnitin does not support dates this far in the future, you may set this, but Turnitin checking will not be available".
2. The event handler for TII which then updates the due dates should be able to determine the latest TII date (or just rely on the TII service reporting an error), and go "meh, I'm turning off TII for this assignment"

This would mean that a user can set a stupidly long deadline, which some institutions may support, especially if they have a single instance for all years (we don't we have an instance *per year*), event queues don't get broken and the plugin in theory should have less work to do.

**Comment by jmcgettrick** [comment]
We'll get this fixed. The best approach from our point of view is probably to warn the user when they are setting the due date that we don't support a date that far in to the future and then limit how far ahead it is set in Turnitin.

**Comment by jmcgettrick** [comment]
We've limited the due date to be - at maximum - a year ahead when the module is recreated in Turnitin. Released in 2015040101.
